### PR TITLE
Improve flexibility of PyFuncModel prediction

### PR DESCRIPTION
**Pull request by skylarbpayne**
## Related Issues/PRs
https://github.com/mlflow/mlflow/issues/694

# Improve flexibility of PyFuncModel prediction

## What changes are proposed in this pull request?

Prior to this change, `PyFuncModel` assumes the implemented class has a `predict` function. This is a common interface, notably used in libraries such as `scikit-learn`. However, there are many use cases where we may want to use the `pyfunc` model flavor (as it is flexible!) while customizing the actual predict function used.

A common use case is illustrated in [issue 694](https://github.com/mlflow/mlflow/issues/694); folks want to use the `predict_proba` function instead of `predict`. However, the resolution in that ticket is effectively to have users write their own custom wrappers.

Given that (as far as I can tell), the pyfunc flavor is intended to "unify" underlying base implementations so you can have a similar interface for potentially dissimilar model formats, it seems prudent to allow folks to customize the predict function used without having to write a custom wrapper. Such functionality is demonstrated in this PR for the `sklearn` model logging, but can possibly be leveraged by other model flavors as well.

## How is this patch tested?

A new test is added which failed before implementation.
Existing tests all pass ensuring backwards compatibility.

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Click the `Details` link on the `Preview docs` check.
2. Find the changed pages / sections and make sure they render correctly.

## Release Notes

Improve flexibility of `PyFuncModel` by allowing predict function override. For example, this can be used to log a `pyfunc` flavor which uses the `scikit-learn` `predict_proba` function instead of `predict`.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/sklearn/__init__.py
Comment (dbczumar) [inline mlflow/sklearn/__init__.py:155]: Nit: Can we use `fn` instead of `func`?

```suggestion
    pyfunc_predict_fn="predict",
```

File: mlflow/sklearn/__init__.py
Comment (dbczumar) [inline mlflow/sklearn/__init__.py:199]: Nit:

```suggestion
    :param pyfunc_predict_fn: The name of the prediction function to use for inference with the pyfunc representation of the resulting MLflow Model; e.g. ``"predict_proba"``.
```

File: mlflow/sklearn/__init__.py
Comment (dbczumar) [inline mlflow/sklearn/__init__.py:375]: Nit: can we make this docstring consistent with the one in `mlflow.sklearn.save_model()`?

File: mlflow/pyfunc/__init__.py
Comment (dbczumar) [inline mlflow/pyfunc/__init__.py:627]:
```suggestion
    def __init__(self, model_meta: Model, model_impl: Any, predict_fn: str = "predict"):
```

File: mlflow/pyfunc/__init__.py
Comment (dbczumar) [inline mlflow/pyfunc/__init__.py:634]: Can we `raise MlflowException` if the predict function isn't available instead of `assert`?

File: tests/sklearn/test_sklearn_model_export.py
Comment (dbczumar) [inline tests/sklearn/test_sklearn_model_export.py:631]: Can we assert that the predictions are equivalent to those produced by `sklearn_logreg_model.predict_proba()`?

File: mlflow/sklearn/__init__.py
Comment (dbczumar) [inline mlflow/sklearn/__init__.py:267]:
```suggestion
            predict_fn=pyfunc_predict_func,
```

**Review by dbczumar — COMMENTED** [review]
@skylarbpayne This is fantastic and long overdue! I just left a few small nits around docs & naming. Should be all set to merge once they're addressed. Thank you for your contribution!

**Comment by WeichenXu123** [comment]
Thanks for the PR !

From my side, I think the ideal solution should be , we add more methods to the PyFuncModel base class, like:

```
class PyFuncModel:

  def predict(..)

  # only available for probability based classifier, otherwise raise unsupported error
  def predict_proba(...)

  # only available for classifier, otherwise raise unsupported error
  def num_classes(...)

```

WDYT ? @dbczumar

**Comment by skylarbpayne** [comment]
> Thanks for the PR !
> 
> From my side, I think the ideal solution should be , we add more methods to the PyFuncModel base class, like:
> 
> ```
> class PyFuncModel:
> 
>   def predict(..)
> 
>   # only available for probability based classifier, otherwise raise unsupported error
>   def predict_proba(...)
> 
>   # only available for classifier, otherwise raise unsupported error
>   def num_classes(...)
> ```
> 
> WDYT ? @dbczumar
> 
> And NOTE that in mlflow `DefaultEvaluator`, it calls `predict` function to compute metrics / generating plots, and assuming "predict" function returning the prediction result, NOT the prob result. If user custom PyFuncModel override "predict" but implement it as "predict_proba", it might cause issues in `DefaultEvaluator`.

Thanks for raising this! How many model flavors have a corresponding `predict_proba`? I think `keras` and `scikit-learn`, but I'm not aware of others. E.g. not `pytorch` to my knowledge. If corresponding `predict_proba` is not common, then it seems already the case that `DefaultEvaluator` might not be broadly usable across flavors anyway, and I wonder whether there's a fix for that or a different `Evaluator` we need?

Regardless, I like the idea of including multiple predict functions (even if I don't like the idea of more work 😅). IMHO, I think such a solution probably warrants a bit more design. In particular, I'm curious about:

- What is the supported set of prediction functions? (e.g. some scikit-learn models also have `decision_function`)
- What are the interfaces of these methods? What further constraints should be expected by users?
- How should users make these available, and in which instances are they available?
- How can we ensure that the runtime loading the model can determine which methods are available? (e.g. with regression models you probably won't have `predict_proba`)

I'll wait for @dbczumar 's thoughts to address any comments as it seems like this thread may change the direction of the solution. Thanks!

File: mlflow/sklearn/__init__.py
Comment (dbczumar) [inline mlflow/sklearn/__init__.py:261]: Can we log a warning if the specified predict function isn't present on the model indicating that the pyfunc model flavor won't be added to the resulting model?

**Review by dbczumar — APPROVED** [review]
LGTM once a small `_logger.warning()` call is introduced (https://github.com/mlflow/mlflow/pull/6631/files#r958772687). Thanks so much, @skylarbpayne !

File: mlflow/sklearn/__init__.py
Comment (skylarbpayne) [reply mlflow/sklearn/__init__.py:261]: Great suggestion! Done!

But maybe another wrench here... 😅. The straightforward way to add a warning log, like I did, will mean that warnings go off for what is possibly legitimate code today (though I'm not familiar with scikit-learn models without a predict function). This code structure makes me think that perhaps what we want is something like:

```
def save_model(..., pyfunc_predict_fn=None):
  ...
  # When
  if not pyfunc_predict_fn and hasattr(sk_model, 'predict'):
    pyfunc_predict_fn = 'predict'

  if pyfunc_predict_fn:
    if not hasattr(sk_model, pyfunc_predict_fn):
      _logger.warning(f'Model was missing {pyfunct_predict_fn} function!')
    else:
      
```

It looks uglier to me (perhaps better way to write it?) But would avoid warnings going off for what is today legitimate code. I'm not familiar enough with scikit-learn models w/o `predict` methods to know if this is worth doing -- so feel free to merge if not. But let me know if it is and I'm happy to comply :)

**Comment by skylarbpayne** [comment]
> LGTM once a small `_logger.warning()` call is introduced (https://github.com/mlflow/mlflow/pull/6631/files#r958772687). Thanks so much, @skylarbpayne !

done, thanks for the swift reviews!

File: mlflow/sklearn/__init__.py
Comment (dbczumar) [reply mlflow/sklearn/__init__.py:262]: I think that adding warning for the existing case(s) where users are logging models that don't define `predict()` is just fine :D. We can always revisit this if necessary. Thanks for considering this!

**Comment by pbower** [comment]
Hi all, 

Could I please confirm how we use this change to use predict_proba but still have it integrate with the serving tools? 

The API reference mentions predict_proba, but for e.g. if we are using predict proba and subsequently thresholding this to produce class predictions, can this integrate with model serving ?

I feel the documentation on this change is potentially unclear, which led to google and then to here.

Please see https://www.mlflow.org/docs/latest/python_api/mlflow.sklearn.html

Thanks

**Comment by pbower** [comment]
Hi, I note this change had 'documentation update required: no" - however should this page be updated? I found it unclear for a use case where I am looking to threshold the probabilities to produce the final decision, in the serving tool.

https://www.mlflow.org/docs/latest/python_api/mlflow.sklearn.html

**Comment by RooieRakkert** [comment]
I've stumbled upon this thread after facing issues using the predict_proba of an (already logged to MLflow) model. 

I've noticed that the  source code for `load_model` [here](https://mlflow.org/docs/latest/_modules/mlflow/pyfunc.html#load_model) tries to find a `predict_fn` within its `conf` (which I assume is the model metadata configuration for the specified FLAVOR_NAME). 

However, I did not find any documentation on where I can actually **_set_** the `predict_fn` during the saving of the model to MLflow. Neither the `mlflow.pyfunc.log_model()` or `mlflow.pyfunc.save_model()` has any argument that seems to hold this configuration.

So my question comes down to; how do I store a custom model to MLflow, whilst also passing the appropriate `predict_fn`, in order to use any other function than the default `.predict()` after loading the model?

**Comment by skylarbpayne** [comment]
Hey @RooieRakkert , thanks for the question.

When I put together this PR, it appears I neglected to provide a straight-forward path for your use case. I think we would similarly need to add `predict_fn` as an argument to `mlflow.pyfunc.log_model`. It is available when you call `mlflow.pyfunc.add_to_model`. So I think this would need to be added to `mlflow.pyfunc.log_model`

I'd be a bit curious to learn a little more about your use case though. In the case that you are using the direct form of `mlflow.pyfunc.log_model`, I would think you have a custom class, where you control which method is "predict" and perhaps have less of a problem, so I may be missing something.

Could you share a little more detail about the object you are trying to log?

**Comment by RooieRakkert** [comment]
Yes you're right. 
Basically our usecase is as follows:

We've created a custom class:


```
class CustomModelWrapper(mlflow.pyfunc.PythonModel):
    """ Wrapper class to store arbitrary model to mlflow""""
    def __init__ (self, model):
        self.model = model
        
    def predict(self, context, model_input: ArrayLike) -> ArrayLike:
        self.model.predict(model_input)
        
    def predict_proba(self, context, model_input:ArrayLike) -> ArrayLike:
      self.model.predict_proba(model_input)
```

We then (after some HyperOpt GridSearch) store our model pipeline (which is an SKlearn pipeline object with an estimator as its final step) with:
```
mlflow.pyfunc.log_model('model_name', python_model=our_model_instance, conda_env=... signature=...)
```

Loading the model is done via:
```
...
model_uri = f"models:/{model_name}/{version}"
model = mlflow.pyfunc.load_model(model_uri=model_uri, dst_path=dst_path)
````


So yes, our current work-around is to overwrite the `.predict()` method of the `CustomModelWrapper` to return `self.model.predict_proba(model_input)`, but I think it would be nicer if we can still have our custom class, which will then have both support for a `.predict()` and `.predict_proba()` method, and then have our MLflow Model specification determine which of the predict- methods we will use by default, after loading up our model.

**Comment by skylarbpayne** [comment]
Ah yes, that makes sense! I am not a maintainer of mlflow, so might make sense to open another issue to assess whether the maintainers are comfortable with the change; I'd be surprised if they weren't. The change isn't particularly "hard", but unfortunately there's multiple places to touch...

In case you or someone else wants to implement it, I think what effectively what needs to happen:

- Add the `predict_fn` argument to [pyfunc.log_model](https://github.com/mlflow/mlflow/blob/master/mlflow/pyfunc/__init__.py#L1492)
- Pass `predict_fn` [through to `Model.log`](https://github.com/mlflow/mlflow/blob/master/mlflow/pyfunc/__init__.py#L1615)

From here, `Model.log` will eventually call back into `pyfunc.save_model`and you'll want to thread the `predict_fn` arg down:

- Add `predict_fn` to [pyfunc.save_model](https://github.com/mlflow/mlflow/blob/master/mlflow/pyfunc/__init__.py#L1312); pay attention to the kwargs here, the maintainers may have particular opinions about how to represent new args (e.g. might want you to just parse out of kwargs instead of making explicit arg?)
- Thread `predict_fn` through the [two](https://github.com/mlflow/mlflow/blob/master/mlflow/pyfunc/__init__.py#L1468) [paths](https://github.com/mlflow/mlflow/blob/master/mlflow/pyfunc/__init__.py#L1479) to save the model, all the way through to the calls to `pyfunc.add_to_model` ... [for example here](https://github.com/mlflow/mlflow/blob/93ee3525205ab217db7261552d3190d4bf02f371/mlflow/pyfunc/model.py#L203)

Hope that's helpful!

**Pull request merged**
